src/main/resources/assets/computercraft/lua/rom/apis/parallel.lua

```diff
@@ -14,15 +13,34 @@ local function create( ... )
     return tCos
 end
 
-local function runUntilLimit( _routines, _limit )
-    local count = #_routines
-    local living = count
+local function runUntilLimit( _routines, _data, _limit )
+    local run, count = true, #_routines
 
     local tFilters = {}
     local eventData = { n = 0 }
+
+    local function checkDead(r, n)
+        local living = count
+        if r and coroutine.status( r ) == "dead" then
+            _routines[n] = nil
+        end
+        for i = 1, count do
+            if _routines[i] == nil then
+                living = living-1
+            end
+        end
+        if living <= _limit then
+            return n
+        end
+    end
+
     while true do
+        local buffer = {}
```

Comment:
I assume you're copying the copying the coroutines here to guard against modifications? Perhaps it would be cleaner to only copy it when the coroutine list is changed - do it within the `:add`/`remove` functions.

It'll mean you'll need to pass in `self` instead of coroutines/data, but that's fine.

Reply:
Do you mean adding a buffer to the group object that `:add/:remove` use, and then when the group is executed, update the `_coroutines` table? I'm confused.

Reply:
So instead of having your remove function be:
```lua
rfuncs[#rfuncs+1] = function()
  self._coroutines[del] = nil
end
```

have something like:

```lua
local current = set[i]
rfuncs[#rfuncs+1] = function()
  local new_routines, n = {}, 0
  for i = 1, self._total do
    local co = self._coroutines[i]
    if co ~= current then
      new_routines[n + 1], n = co, n + 1
    else
      self._coroutines[i] = nil
    end
  end
  self._coroutines = new_routines
  self._total = n
end
```

Effectively, removing an item does two things:
 - Delete it from the current set, but don't remove it - we don't want to shrink the list.
 - Replace `_coroutines`/`_total` with a new collection with the item removed.

This means here you just need to this at each iteration - no need to copy the whole `_routines` list:

```lua
local buffer, count = self._coroutines, self._total
```

Reply:
Of course, you'll need to pass in the `self` object instead of `_routines`/`_limit`/`_data` - I don't think this is an issue if you do the any/all flag as mentioned elsewhere.

Reply:
Right so I've changed how removal works. Since it's now a function of the group 'class' the `add` function returns an ID that is then used later on to remove the thread. Take a look at the latest commit and give me a shout with what should be done.